0024-pre-shared-key-v2/0024-pre-shared-key-v2.md

```diff
@@ -0,0 +1,287 @@
+7. Receiver creates a [Payment Chunk Response](#payment-chunk-response) with the same Payment ID and Sequence number. The Payment Amount and Chunk Amount are set to the amount that arrived on the incoming transfer. The receiver then [encrypts](#encryption-envelope) the data.
+8. Receiver generates fulfillment as described [below](#fulfillment-generation) and fulfills the incoming transfer with the fulfillment and the encrypted response data.
+9. Sender is notified of the outgoing transfer being fulfilled and can use the decrypted response to determine the amount that arrived at the destination.
+
+### Chunked Payment
+
+1. Sender creates a packet with Type 0, a random Payment ID, a Sequence number of 0 and a Chunk Amount of 0. In the case of a fixed destination amount, sender sets the Payment Amount to the destination amount. In the case of a fixed source amount, sender sets the Payment Amount to 2^64 - 1 (the maximum unsigned 64-bit integer, or 18446744073709551615). Sender [encrypts the request](#encryption-envelope).
+2. Sender generates the payment condition as described [below](#fulfillable-condition).
+3. Sender sends a payment with the condition from step 2., the encrypted request from step 1. as the data, and an arbitrary source amount.
+4. Receiver is notified of the incoming transfer and decrypts the data.
+5. Receiver checks that the incoming transfer amount is greater than or equal to the Chunk Amount and that the total that has arrived thus far does not exceed the Payment Amount.
+6. Receiver creates a response packet with the Type 2, the same Payment ID, Sequence. The Payment Amount is set to the total the receiver has received and fulfilled thus far (including the current chunk). Receiver sets the Chunk Amount to the amount that arrived on the incoming transfer. Receiver [encrypts the response](#encryption-envelope).
+7. Receiver generates fulfillment as described [below](#fulfillment-generation) and fulfills the incoming transfer.
+8. Sender is notified of the outgoing transfer being fulfilled and decrypts the attached data.
+9. Sender determines the exchange rate by dividing the Chunk Amount from the receiver by the chunk's source amount.
+10. Sender repeats step 1-8 with the following modifications, until the fixed source or destination amount has been sent or delivered, respectively:
+  - Sender MUST create a new PSK data packet for each chunk with the same Payment ID and an incremented Sequence number
+  - Sender MAY specify the minimum Chunk Amount the receiver should accept for each transfer based on the calculated path exchange rate
+  - Sender MAY adjust the transfer amount upwards or downwards in response to successful and unsuccessful payments, or to deliver a fixed destination amount exactly
+  - Sender SHOULD set the Type to 1 on the last chunk
```

Comment:
What if the sender didn't set the type to 1, but the chunk they sent completes the payment's destination amount. Would the receiver reject subsequent payments? Do they have a way of telling the sender that they've completed the payment already?

What if the sender sets the type to 1, but the destination amount isn't complete yet? Does the receiver just take it to mean that the sender won't pay the rest of what they owe?

Reply:
> What if the sender didn't set the type to 1, but the chunk they sent completes the payment's destination amount. Would the receiver reject subsequent payments?

Yes, the receiver should reject further chunks once the destination amount specified has arrived

> What if the sender sets the type to 1, but the destination amount isn't complete yet? 

The receiver should assume that means the sender won't send any more. I think the PSK implementation would need to turn it over to the Application Layer protocol at that point to figure out what to do